File: Common/Assert.vb
Namespace Common

    Public Class Assert

        <DebuggerHidden()>
        Public Shared Sub NonNull(Of T As Class)(ByVal O As T, Optional ByVal Name As String = Nothing)
            If O Is Nothing Then
                If Name IsNot Nothing Then
                    Fail(String.Format("Object reference '{0}' not set to an instance of an object.", Name))
                Else
                    Fail("Object reference not set to an instance of an object.")
                End If
            End If
        End Sub

        <DebuggerHidden()>
        Public Shared Sub [True](ByVal T As Boolean, Optional ByVal Message As String = Nothing)
            If Not T Then
                Fail(Message)
            End If
        End Sub

        <DebuggerHidden()>
        Public Shared Sub [False](ByVal T As Boolean, Optional ByVal Message As String = Nothing)
            If T Then
                Fail(Message)
            End If
        End Sub

        <DebuggerHidden()>
        Public Shared Sub Fail(Optional ByVal Message As String = Nothing)
            Throw New AssertionException(Message)
        End Sub

    End Class

End Namespace

File: Common/Assert.cs
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;

namespace Ks.Common
{
    public class Assert
    {
        [DebuggerHidden()]
        public static void NonNull<T>([NotNull] T O, string Name = null) where T : class
        {
            if (O == null)
            {
                if (Name != null)
                {
                    Fail(string.Format("Object reference '{0}' not set to an instance of an object.", Name));
                }
                else
                {
                    Fail("Object reference not set to an instance of an object.");
                }
            }
        }

        [DebuggerHidden()]
        public static void True([DoesNotReturnIf(false)] bool T, string Message = null)
        {
            if (!T)
            {
                Fail(Message);
            }
        }

        [DebuggerHidden()]
        public static void False([DoesNotReturnIf(true)] bool T, string Message = null)
        {
            if (T)
            {
                Fail(Message);
            }
        }

        [DebuggerHidden()]
        [DoesNotReturn]
        public static AssertionException Fail(string Message = null)
        {
            throw new AssertionException(Message);
        }
    }
}
